### PR TITLE
prov/sockets: set address size correctly in fi_getname/fi_getpeer

### LINKED ISSUE #985: sockets: fi_getname()/fi_getpeer() set addrlen incorrectly
The man page for `fi_getname()` and `fi_getpeer()` states: 
> On output, addrlen is set to the size of the buffer needed to store the address, which may be larger than the input value.

The sockets provider instead sets addrlen to the number of bytes actually stored.  The sockets provider should be updated to match the description in the man page.

@jithinjosepkl

### PR DESCRIPTION
**Pull request by jithinjosepkl**
- Fixes #985 

@shefty, @shantonu  - please review.

### CODE REVIEW

File: prov/sockets/src/sock_ep_msg.c [hunk 1]
```diff
@@ -244,13 +244,12 @@ static int sock_ep_cm_getname(fid_t fid, void *addr, size_t *addrlen)
 	struct sock_ep *sock_ep = NULL;
 	struct sock_pep *sock_pep = NULL;
 
-	if (*addrlen == 0) {
+	if (*addrlen < sizeof(struct sockaddr_in)) {
```
Comment (shefty) [inline prov/sockets/src/sock_ep_msg.c]: The original code is correct.  The address should be truncated if addrlen is too small.

File: prov/sockets/src/sock_ep_msg.c [hunk 2]
```diff
@@ -244,13 +244,12 @@ static int sock_ep_cm_getname(fid_t fid, void *addr, size_t *addrlen)
...
 		*addrlen = sizeof(struct sockaddr_in);
 		return -FI_ETOOSMALL;
 	}
 
-	*addrlen = MIN(*addrlen, sizeof(struct sockaddr_in));
-
+	*addrlen = sizeof(struct sockaddr_in);
```
Comment (shefty) [inline prov/sockets/src/sock_ep_msg.c]: You want to keep the MIN type of call.  I would add a new var 'buflen', set buflen = MIN(), and use that in the memcpy routines.  Then always set addrlen = sizeof sockaddr_in.

Comment (jithinjosepkl) [inline prov/sockets/src/sock_ep_msg.c]: Ah, got it. So app would find out truncation by seeing a larger len in addrlen field.  Will update the PR.

**Comment by jithinjosepkl** [comment]
@shefty - I have updated the code.

File: prov/sockets/src/sock_ep_msg.c [hunk 3]
```diff
@@ -243,29 +243,31 @@ static int sock_ep_cm_getname(fid_t fid, void *addr, size_t *addrlen)
...
+	size_t len;
 
 	if (*addrlen == 0) {
 		*addrlen = sizeof(struct sockaddr_in);
 		return -FI_ETOOSMALL;
 	}
 
-	*addrlen = MIN(*addrlen, sizeof(struct sockaddr_in));
-
+	len = MIN(*addrlen, sizeof(struct sockaddr_in));
```
Comment (shefty) [inline prov/sockets/src/sock_ep_msg.c]: This function looks correct to me.  As an alternative, you could do this to have a single exit point:

len = MIN(...)
if (len ==)
   goto out;   /* jumps to setting addrlen below */

File: prov/sockets/src/sock_ep_msg.c [hunk 4]
```diff
@@ -378,9 +381,9 @@ static int sock_ep_cm_getpeer(struct fid_ep *ep, void *addr, size_t *addrlen)
 	}
 
 	sock_ep = container_of(ep, struct sock_ep, ep);
-	*addrlen = MIN(*addrlen, sizeof(struct sockaddr_in));
-	memcpy(addr, sock_ep->dest_addr, *addrlen);
-	return (*addrlen == sizeof(struct sockaddr_in)) ? 0 : -FI_ETOOSMALL;
+	len = MIN(*addrlen, sizeof(struct sockaddr_in));
+	memcpy(addr, sock_ep->dest_addr, len);
+	return (len == sizeof(struct sockaddr_in)) ? 0 : -FI_ETOOSMALL;
```
Comment (shefty) [inline prov/sockets/src/sock_ep_msg.c]: This function does not set addrlen.  Also, does memcpy handle len = 0?  If not, you'll need something like what you have for getname.  If so, then getname could be updated to match this flow.

**Comment by jithinjosepkl** [comment]
Need more coffee i guess :) 

Reg. 0 byte memcpy - I tried a simple pgm with memcpy (len = 0), and it was fine. So, i removed the first check for len == 0. Please take a look.

**Comment by shefty** [comment]
Thanks - will you also merge into v1.1.x?

**Comment by jithinjosepkl** [comment]
Yes, done.

**Pull request merged**